rskj-core/src/main/java/co/rsk/db/importer/BootstrapImporter.java

```diff
@@ -102,10 +81,58 @@ private static void insertBlocks(BlockStore blockStore,
             RLPList blockData = RLP.decodeList(element.getRLPData());
             RLPList tuple = RLP.decodeList(blockData.getRLPData());
             Block block = blockFactory.decodeBlock(tuple.get(0).getRLPData());
-            BlockDifficulty blockDifficulty = new BlockDifficulty(new BigInteger(tuple.get(1).getRLPData()));
+            BlockDifficulty blockDifficulty = new BlockDifficulty(new BigInteger(Objects.requireNonNull(tuple.get(1).getRLPData())));
             blockStore.saveBlock(block, blockDifficulty, true);
         }
 
         blockStore.flush();
     }
+
+    private static void prepareState(Queue<byte[]> nodeDataQueue, HashMapDB hashMapDB, RLPElement rlpElement) {
+        Queue<RLPElement> nodeListQueue = decodeQueue(rlpElement.getRLPData());
+
+        prepareNodeData(nodeDataQueue, hashMapDB, RLP.decodeList(Objects.requireNonNull(nodeListQueue.poll()).getRLPData()));
+        prepareNodeValues(hashMapDB, RLP.decodeList(Objects.requireNonNull(nodeListQueue.poll()).getRLPData()));
+    }
+
+    private static void prepareNodeData(Queue<byte[]> rlpDataQueue, HashMapDB hashMapDB, RLPList nodesData) {
+        int size = nodesData.size();
+        for (int k = 0; k < size; k++) {
+            RLPElement element = nodesData.get(k);
+            byte[] rlpData = Objects.requireNonNull(element.getRLPData());
+
+            hashMapDB.put(Keccak256Helper.keccak256(rlpData), rlpData);
+            rlpDataQueue.add(rlpData);
+        }
+    }
+
+    private static void prepareNodeValues(HashMapDB hashMapDB, RLPList nodeValues) {
+        int size = nodeValues.size();
+        for (int k = 0; k < size; k++) {
+            RLPElement element = nodeValues.get(k);
+            byte[] rlpData = element.getRLPData();
+            hashMapDB.put(Keccak256Helper.keccak256(rlpData), rlpData);
+        }
+    }
+
+    private static void insertState(TrieStore destinationTrieStore, Queue<byte[]> nodeDataQueue, HashMapDB hashMapDB) {
+        TrieStoreImpl fakeStore = new TrieStoreImpl(hashMapDB);
+        for (byte[] nodeData = nodeDataQueue.poll(); nodeData != null; nodeData = nodeDataQueue.poll()) {
+            Trie trie = Trie.fromMessage(nodeData, fakeStore);
+
+            destinationTrieStore.save(trie);
+        }
+    }
+
+    private static Queue<RLPElement> decodeQueue(byte[] data) {
+        RLPList rlpList = RLP.decodeList(data);
+        int size = rlpList.size();
+
+        LinkedList<RLPElement> result = new LinkedList<>();
```

Comment:
Can we make this variable of type `Queue<RLPElement>`? Maybe I'm missing something

Reply:
Not sure I got the point here. `Queue` is an interface and `LinkedList` is one of its implementations.

Reply:
ahh.. I see)

Reply:
Done